js/net.ts

```diff
@@ -0,0 +1,151 @@
+
+import { ReadResult, Reader, Writer, Closer } from "./io";
+import * as fbs from "gen/msg_generated";
+import { assert, notImplemented } from "./util";
+import * as dispatch from "./dispatch";
+import { flatbuffers } from "flatbuffers";
+import { read, write, close } from "./files";
+
+export type Network = "tcp";
+// TODO support other types:
+// export type Network = "tcp" | "tcp4" | "tcp6" | "unix" | "unixpacket";
+
+// TODO Support finding network from Addr, see https://golang.org/pkg/net/#Addr
+export type Addr = string;
+
+/** A Listener is a generic network listener for stream-oriented protocols. */
+export interface Listener {
+  /** accept() waits for and returns the next connection to the Listener. */
+  accept(): Promise<Conn>;
+
+  /** Close closes the listener.
+   * Any pending accept promises will be rejected with errors.
+   */
+  close(): void;
+
+  addr(): Addr;
+}
+
+class ListenerImpl implements Listener {
+  constructor(readonly fd: number) {}
+
+  async accept(): Promise<Conn> {
+    const builder = new flatbuffers.Builder();
+    fbs.Accept.startAccept(builder);
+    fbs.Accept.addRid(builder, this.fd);
+    const msg = fbs.Accept.endAccept(builder);
+    const baseRes = await dispatch.sendAsync(builder, fbs.Any.Accept, msg);
+    assert(baseRes != null);
+    assert(fbs.Any.NewConn === baseRes!.msgType());
+    const res = new fbs.NewConn();
+    assert(baseRes!.msg(res) != null);
+    return new ConnImpl(res.rid(), res.remoteAddr()!, res.localAddr()!);
+  }
+
+  close(): void {
+    close(this.fd);
+  }
+
+  addr(): Addr {
+    return notImplemented();
+  }
+}
+
+export interface Conn extends Reader, Writer, Closer {
+  localAddr: string;
+  remoteAddr: string;
+}
+
+class ConnImpl implements Conn {
+  constructor(
+    readonly fd: number,
+    readonly remoteAddr: string,
+    readonly localAddr: string
+  ) {}
+
+  write(p: ArrayBufferView): Promise<number> {
+    return write(this.fd, p);
+  }
+
+  read(p: ArrayBufferView): Promise<ReadResult> {
+    return read(this.fd, p);
+  }
+
+  close(): void {
+    close(this.fd);
+  }
+}
+
+/** Listen announces on the local network address.
+ *
+ * The network must be "tcp", "tcp4", "tcp6", "unix" or "unixpacket".
+ *
+ * For TCP networks, if the host in the address parameter is empty or a literal
+ * unspecified IP address, Listen listens on all available unicast and anycast
+ * IP addresses of the local system. To only use IPv4, use network "tcp4". The
+ * address can use a host name, but this is not recommended, because it will
+ * create a listener for at most one of the host's IP addresses. If the port in
+ * the address parameter is empty or "0", as in "127.0.0.1:" or "[::1]:0", a
+ * port number is automatically chosen. The Addr method of Listener can be used
+ * to discover the chosen port.
+ *
+ * See dial() for a description of the network and address parameters.
+ */
+export function listen(network: Network, address: string): Listener {
```

Comment:
I find "network" very confusing. Why not "protocol" like everybody else and their cat?

Reply:
It comes from Golang https://golang.org/pkg/net/#Listen

Reply:
I see. Still a bad name. 
"unix packet" has nothing to do with network.
Soon well have "websocket" which is also not a network.